cachi2/core/package_managers/bundler/parser.py

```diff
@@ -0,0 +1,157 @@
+    Base class for gem metadata.
+
+    Attributes:
+        name:       The name of the gem.
+        version:    The version of the gem.
+    """
+
+    name: str
+    version: str
+
+
+class GemDependency(_GemMetadata):
+    """
+    Represents a gem dependency.
+
+    Attributes:
+        source:     The source URL of the gem.
+        checksum:   The checksum of the gem.
+    """
+
+    source: str
+    checksum: Optional[str] = None
+
+    @property
+    def purl(self) -> str:
+        """Get purl for this gem dependency."""
+        return NotImplemented
+
+
+class GitDependency(_GemMetadata):
+    """
+    Represents a git dependency.
+
+    Attributes:
+        url:        The URL of the git repository.
+        revision:   Commit hash.
+        branch:     The branch of the git repository.
+    """
+
+    url: str
+    revision: str
+    branch: Optional[str] = None
+
+    @property
+    def purl(self) -> str:
+        """Get purl for this git dependency."""
+        return NotImplemented
+
+    @pydantic.field_validator("url")
+    def validate_git_url(cls, url: str) -> str:
+        """Validate that the git URL is a valid git URL."""
+        parsed_url = urlparse(url)
+        if parsed_url.scheme != "https":
+            raise ValueError("GIT dependencies must use HTTPS protocol")
+
+        # TODO: more validation
+
+        return url
+
+    @pydantic.field_validator("revision")
+    def validate_git_revision(cls, revision: str) -> str:
+        """Validate that the git revision is a valid git ref."""
+        if not re.match(r"^[a-fA-F0-9]{40}$", revision):
+            raise ValueError("GIT dependencies must have a valid revision")
+
+        return revision
+
+
+class PathDependency(_GemMetadata):
+    """
+    Represents a path dependency.
+
+    Attributes:
+        path:       Subpath from package root.
+    """
+
+    path: str
+
+    @property
+    def purl(self) -> str:
+        """Get purl for this path dependency."""
+        return NotImplemented
+
+
+BundlerDependency = Union[GemDependency, GitDependency, PathDependency]
+ParseResult = list[BundlerDependency]
+
+
+def parse_lockfile(package_dir: RootedPath) -> ParseResult:
+    """Parse a Gemfile.lock file and return a list of dependencies."""
+    if (
+        not package_dir.join_within_root(GEMFILE).path.exists()
+        or not package_dir.join_within_root(GEMFILE_LOCK).path.exists()
+    ):
+        reason = "Gemfile and Gemfile.lock must be present in the package directory"
+        solution = (
+            "Run `bundle init` to generate the Gemfile.\n"
+            "Run `bundle lock` to generate the Gemfile.lock."
+        )
+        raise PackageRejected(reason=reason, solution=solution)
+
+    bundler_package = files("cachi2.core.package_managers.bundler")
+    lockfile_parser_script = str(bundler_package.joinpath("scripts").joinpath("lockfile_parser.rb"))
+    try:
+        output = run_cmd(cmd=[lockfile_parser_script], params={"cwd": package_dir.path})
```

Comment:
I like how bundler correctly dumps warnings to stderr which in the default configuration of subprocess.run gets ignored and hence doesn't crash the json parser: 
```
Warning: the running version of Bundler (2.2.33) is older than the version that created the lockfile (2.5.10). We suggest you to upgrade to the version that created the lockfile by running `gem install bundler:2.5.10`.
```